proposals/0010-diet-clients.txt

```diff
@@ -0,0 +1,205 @@
+state transition or a double confirmation attack on the network.
+
+## Specification
+
+### Overview
+
+1. User observes a confirmation for a TX at slot N. Confirmation
+contains 2/3+ votes for slot N, and a path from the transaction to
+the bankhash signed by all the votes.
+
+2. Concurrently with 1, user requests votes for slot N from M random
+validators.  Votes returned may be for later slots, and would contain
+a merkle path to the requested slot's bankhash. M should be large
+enough to represent a small but significant percentage of the stake
+above X%. X should  be configured by the client as the desired
+threshold for confidence.
+
+3. If all the responses are the same, the user is done. 
+
+4. If X% or more of staked nodes respond that they do not have block,
+the user starts sampling nodes at random for missing shreds and
+submitting the shreds to the validators with the missing block.
+
+    a. This process may need to run recursively until the fork block
+    is repaired.
+
+    b. If shreds cannot be recovered, a fault has occurred. User
+    shouldn't trust the confirmation.
+
+5. If X% or more of the staked nodes respond with different
+bankhash, a fault has occurred on the network. User shouldn't trust
+the confirmation. User should then submit the conflicting votes to
+M random validators. User shouldn't trust any confirmations from
+the network.
+
+6. if X% or more of the staked nodes do not respond in T seconds a
+faulty majority may have partitioned the network and the client.
+Client should ignore the confirmation, manually confirm those nodes
+are down, and remove them from the local view of quorum for this
+transaction and try again. X should be configured large enough that
+this doesn't happen more then once a year. Clients can automatically
+remove nodes from the local view of the quorum which have not voted
+for K slots.
+
+### Confirmation Proof
+
+#### BankHash
+
+BankHash = Merkle(ShredHash, BlockHash, StateHash, Prev BankHash)
+
+#### ShredHash: Data availability merkle tree
+
+This tree is necessary for step 4 in the protocol. If enough nodes
+respond that they do not have the data, the client should try to
+repair the data for them. This prevents the faulty majority from
+withholding the data from the rest of the network, so the rest can
+identify an invalid state transition or a double confirmation.
+
+Turbine, solana's block propagation protocol, transmits blocks in
+small chunks called shreds. Shreds are transmitted in batches.
+Each batch includes J data shreds and K erasure codes.
+
+Upon executing the block, the validator should compute the merkle
+root of all the shreds, including the erasure shreds. The merkle
+root should be added to the BankHash, which is signed by validators
+when voting.
+
+When sampling for shreds, the user knows that the shreds are for
+the proposed confirmed block because each shred has a merkle path
+to the bankhash that was signed by all the validators.
```

Comment:
Don't the validators sign the merkle root generated by merkelizing all the shreds https://github.com/solana-labs/solana/blob/9db25655f72dc3cd4362b9ef077f09054573ccee/ledger/src/shred/merkle.rs#L1119
which is diff from the bank hash https://github.com/solana-labs/solana/blob/283a2b15402867baeb0735acda1cdc8632afd400/runtime/src/bank.rs#L6714-L6719

Reply:
Yea, this would need to change.  We need the validators to sign the merkle root as well.